### ISSUE TITLE
Variadic Terminals

### ISSUE THREAD

**Issue by knpwrs** [Hardmath123/nearley#236]
I'm very new to the world of parser generators. Is there a way I can define terminal strings based on input to a parser? For instance, consider the following HL7 message:

```
MSH|^~\&|FOO|BAR|BAZ
OBX|bear|banana^apple
```

At first glance it looks like some sort of CSV-type format. The tricky part is that the terminals are actually defined by the five characters immediately following MSH. The following is the same message expressed with different control characters:

```
MSH$%*#@$FOO$BAR$BAZ
OBX$bear$banana%apple
```

HL7 messages _typically_ use `|^~\&`, but they don't have to. Is there a way to support this with nearley?

**Comment by tjvr** [comment]
Most parser generators don't do this.

Depending on your format, it might be possible to define the grammar rules on-the-fly, and Nearley (unlike most parsing tools) totally lets you do this. Use `nearley.Grammar.fromCompiled(require('my-grammar.js'))` to import your grammar after compiling it with `nearleyc`, and then add additional `nearley.Rule`s to the grammar to define the non-terminals. :-)

Failing that, you might have to skip the parser generator and implement a recursive descent parser by hand... Presumably you've looked for existing HL7 parsers?

@Hardmath123 What do you think?

**Comment by knpwrs** [comment]
There are HL7 parsers available, but for my purposes I need index or row/column information for the parsed data. I saw that I could make something like this with nearley. I actually got something working with [`moo`][http://npm.im.moo] though obviously not a full parser, just a lexer (the way I got it working there was to parse the header and then generate a lexer with the extracted information).

I'd still like to get something working with Nearley, if possible, since it can bring me the rest of the way in parsing the data.

Nearley is able to use `moo` as a lexer, right? Can I specify my `moo` lexer after compilation with `nearleyc` and use that?
